src/dataset/create_dataset.py

```diff
@@ -19,240 +20,141 @@
+            if (self.custom_sources is not None) and (entry['source'] not in self.custom_sources):
+                continue
+
+            # raise error if the entry does not contain the required metadata
+            self.contains_required_metadata(entry)
+
+            #if the text is too short, ignore this text
+            if len(entry['text']) < 500:
+                continue
+
+
+            """Checks are done, so we construct the metadata."""
+
+            # Get id, source, title, text, url, date_published, authors, and summary
+            entry_id: str = entry['id']
+            source: str = entry['source']
+            title: str = entry['title']
+            text: str = entry['text']
+            url: str = entry['url']
+            date_published: str = entry['date_published']
+            authors: list = entry['authors']
+            # summary is ignored for now, since most sources lack one. TODO: add summary. see self.metadata code as well.
+            
+            # Get signature
+            if authors:
+                signature = f"Title: {title}, Authors: {get_authors_str(authors)}"
+            else:
+                signature = f"Title: {title}"
+
+            # We use the text_splitter to get the chunks from the entry, 
+            # and we add a metadata element for each new chunk we add to the dataset.
+            chunks = text_splitter.split(text, signature)
+            num_chunks = len(chunks)
+            
+            for i in range(num_chunks):
+                self.metadata.append({
+                    'id': f"{entry_id}_{str(i+1).zfill(6)}",
+                    'entry_id': entry_id,
+                    'source': source,
+                    'title': title,
+                    'text': chunks[i],
+                    'url': url,
+                    'date_published': date_published,
+                    'authors': authors
+                })
+            
+            # Update counts
+            self.entries_per_source_count[entry['source']] += 1
+            self.total_counts['entries'] += 1
+            self.total_counts['chars'] += len(text)
+            self.total_counts['words'] += len(text.split())
+            self.total_counts['sentences'] += len(split_into_sentences(text))
+            self.total_counts['chunks'] += len(chunks)
+
+    def show_stats(self):
+        print(f'Number of entries by source: {self.entries_per_source_count}')
+        print(f'Total entries count: {self.total_counts["entries"]}')
+        print(f'Total character count: {self.total_counts["chars"]}')
+        print(f'Total word count: {self.total_counts["words"]}')
+        print(f'Total sentence count: {self.total_counts["sentences"]}')
+        print(f'Total chunk count: {self.total_counts["chunks"]}')
 
     def get_embeddings(self):
         def get_embeddings_at_index(texts: str, batch_idx: int, batch_size: int = 200): # int, np.ndarray
```

Comment:
is it possible to do the assignments in the for loop directly via numpy? I have a vauge impression that can massively speed things up. Though I might be thinking of something else...

Reply:
Yeah I didn't really touch get_embeddings, it's a bit confusing and complicated, but it can be improved for sure I'll test things later